packages/rule-https-only/src/rule.ts

```diff
@@ -0,0 +1,69 @@
+/**
+ * @fileoverview sonarwhal rule that checks if your site is using HTTPS and if it has mixed content.
+ */
+
+import { Category } from 'sonarwhal/dist/src/lib/enums/category';
+import { RuleContext } from 'sonarwhal/dist/src/lib/rule-context';
+
+import { IRule, RuleMetadata, Event } from 'sonarwhal/dist/src/lib/types';
+import { debug as d } from 'sonarwhal/dist/src/lib/utils/debug';
+import { RuleScope } from 'sonarwhal/dist/src/lib/enums/rulescope';
+import { isHTTPS } from 'sonarwhal/dist/src/lib/utils/misc';
+
+const debug: debug.IDebugger = d(__filename);
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+export default class HttpsOnlyRule implements IRule {
+    private targetIsServedOverHTTPS: boolean = false;
+
+    public static readonly meta: RuleMetadata = {
+        docs: {
+            category: Category.security,
+            description: `sonarwhal rule that checks if your site is using HTTPS and if it has mixed content.`
+        },
+        id: 'https-only',
+        schema: [],
+        scope: RuleScope.site
+    }
+
+    public constructor(context: RuleContext) {
+
+        const validateTarget = async (event: Event): Promise<void> => {
+            const { resource } = event;
+
+            if (!isHTTPS(resource)) {
+                debug('HTTPS no detected');
+
+                await context.report(resource, null, 'The site should be HTTPS');
+
+                return;
+            }
+
+            debug('HTTPS detected');
+
+            this.targetIsServedOverHTTPS = true;
+
+            return;
+        };
+
+        const validate = async (event: Event) => {
+            if (!this.targetIsServedOverHTTPS) {
+                return;
+            }
+
+            const { resource } = event;
+
+            if (!isHTTPS(resource)) {
+                await context.report(resource, null, 'Should be served over HTTPS');
+            }
+        };
+
+        context.on('fetch::start::target', validateTarget);
```

Comment:
I think we should subscribe to the `fetch::end::target` because there could be a redirect from HTTP to HTTPS and I don't think we should penalize that. If there are redirects for resources we should fail those.

Reply:
Also I would subscribe to all `fetch::end`. There could be a weird redirect case like HTTPS -->HTTP that is not detected here so we should check that all the `hops` are on HTTPS.

Reply:
There is no `fetch::end::target` event, but probably it would be nice to have it.

> Also I would subscribe to all fetch::end. There could be a weird redirect case like HTTPS -->HTTP that is not detected here so we should check that all the hops are on HTTPS.

I think that starting a fetch to `http` can be also a problem, I think that an http petition is vulnerable even if it redirect to a `https`. But I'm not sure...

Reply:
Yes, but we already have some rules that touch on this like [STS](https://sonarwhal.com/docs/user-guide/rules/rule-strict-transport-security/) so I think we should be ok with that. Otherwise you will need to know reuse a lot of the code in there, verify if the site is in the list, etc.

Reply:
Then, I need to add the event `fetch::end::target` and change the events to `fetch::end::*`, right?

Reply:
Yes, that's what I think. And check the `hops` property of `response`

Reply:
ok

Reply:
Only the target